envoy/http/filter.h

```diff
@@ -868,11 +874,16 @@ class StreamEncoderFilterCallbacks : public virtual StreamFilterCallbacks {
    * @param modify_headers supplies an optional callback function that can modify the
    *                       response headers.
    * @param grpc_status the gRPC status code to override the httpToGrpcStatus mapping with.
+   * @param grpc_error_details the gRPC status error details. It will be seralized and encoded in
+   * grpc-status-details-bin header. Please be aware, inside `grpc_error_details`, only the
+   * `details` field will be effective and the `code`/`message` fields will be overwritten by the
+   * provided parameters.
    * @param details a string detailing why this local reply was sent.
    */
   virtual void sendLocalReply(Code response_code, absl::string_view body_text,
                               std::function<void(ResponseHeaderMap& headers)> modify_headers,
                               const absl::optional<Grpc::Status::GrpcStatus> grpc_status,
+                              std::unique_ptr<::google::rpc::Status> grpc_error_details,
```

Comment:
to be consistent,  we could use "const absl::optional"

Reply:
Disagree. `std::optional` will create a copy. go/totw/163

The `grpc_status` param is an enum so it's OK to create copy. But not for `google::rpc::Status`, it may be a big proto.

Reply:
Anyways, I think parameter should be type `const ::google::rpc::Status*`. Why do we need `unique_ptr`? Caller can maintain ownership of the original object, no need to transfer ownership for function call.

Reply:
Yeah, I don't want to copy so I uses the unique pointer.

If we go with `const ::google::rpc::Status*`, the caller is responsible to set `code/message` field. Of course, we can use `::google::rpc::Status*` but the callee will still set  `code/message` field, which will surprise caller. Anyway, I think the caller doesn't need `::google::rpc::Status` anymore so no need to hold it. WDYT?

Reply:
Right, `::google::rpc::Status*` would be a better choice over `const`

Usually we prefer to use raw pointer over `unique_ptr` if there is no need for ownership transfer.
With raw pointer, caller can allocate on the stack easily.

```
Status s;
sendLocalReply(..., &s, ...);
```

Up to you since you already setup everything as `unique_ptr`.

Reply:
Shouldn't be too much difference between
```
Status s;
// update s
sendLocalReply(..., &s, ...);
```
```
std::unique_ptr<Status> s = std::make_unique<Status>();
// update s
sendLocalReply(..., std::move(s), ...);
```

Still, I don't want caller to hold the object, because `Status` will be modified. It will either surprise callers or make callers depend on the modification from `sendLocalReply`. Anyway, we shouldn't allow callers to access it again, which will be enforced by `unique_ptr` and `move`.

Reply:
Ok, sounds good